more_itertools/more.py

```diff
@@ -4521,3 +4522,40 @@ def partial_product(*iterables):
     for i, it in enumerate(iterators):
         for prod[i] in it:
             yield tuple(prod)
+
+
+def _discriminate_iterable(idx, iterable):
```

Comment:
Had to break this out into a function to avoid issues with name binding inside of generators